### ISSUE TITLE
Improve the catalog endpoint format

### ISSUE THREAD

**Issue by arschles** [gomods/athens#1186]
**Is your feature request related to a problem? Please describe.**

Right now, utilities that want to use the `/catalog` endpoint often need to aggregate each module's versions into a list. For example, the `/catalog` endpoint returns this:

```json
{"modules": [
  {
    "module":"github.com/arschles/assert",
    "version": "v1.0.0"
  },
  {
    "module": "github.com/arschles/assert",
    "version: "v1.1.0"
  }
}]
```

Many programs will find it more convenient to get a response similar to this: 

```json
{
    "modules": {
        "github.com/arschles/assert": ["v1.0.0", "v1.1.0"]
    }
}
```

**Describe the solution you'd like**

I'd like us to add a V2 of the catalog API that responds this way. This could possibly be exposed as a `/catalog/v2` endpoint.

**Describe alternatives you've considered**

Clients can always aggregate in their own code

**Additional context**

There will be non-trivial work on the storage systems to achieve this

**Comment by MaruHyl** [comment]
@arschles It looks simple, can I try it?

**Comment by marpio** [comment]
@MaruHyl sure, it's yours!
One thing you should keep in mind is that we have paging in place currently. This means that the last record we're getting from the storage doesn't have to be the last version of that module.
page1:
mod1@v1.00
mod1@v1.0.0
mod2@v2.0.0
Page2:
mod2@v2.1.0
etc.
So we can't just simply group by module the page records.

**Comment by MaruHyl** [comment]
@marpio Thanks!
I will use `module` as the key for sorting and paging. For example, pagesize is 3, then the result is m1(v1,v2..), m2(v1,v2..),m3(v1,v2..).

**Comment by MaruHyl** [comment]
@marpio 
In `fs.CatalogV2`, the token will be `module`, not `module|version`.For example, pagesize is 3, then the result is m1(v1,v2..), m2(v1,v2..),m3(v1,v2..), the next token is m4
In `mongo.CatalogV2`, the token will be `module`, not `ID`, and the next token is m4 too.
...

**Comment by manugupt1** [comment]
Hi,

My preference is not to sort as the order is not guaranteed by any storage
providers. The two storages may not necessarily be the same.


Cheers,

Manu Gupta

**Comment by marpio** [comment]
@MaruHyl
Sorry for not being clear, I was thinking about the object storages (S3,GCS etc.)
Since we don't know the number of versions for each module we will need to `List` more objects that we actually need to satisfy the pageSize number of modules. 
Now we can't simply return the last continuation token returned be the storage b/c we would skip some records.
Just something we should think of while implementing this.

**Comment by MaruHyl** [comment]
@marpio @manugupt1 Thanks for advises!

**Comment by 5onic** [comment]
@MaruHyl Hey, do you have some progress on it? if you don't have time, I'll take care of it:)

**Comment by MaruHyl** [comment]
@5onic sure, you can take it.

**Comment by 5onic** [comment]
@MaruHyl Got it.

So, I'll work on this.

**Comment by marpio** [comment]
@arschles @5onic 
What do you think about supporting the https://index.golang.org/index format instead?

**Comment by 5onic** [comment]
@marpio @arschles  Yes, it would be cool to use the format, that you mentioned. And I'm pretty sure that it will be helpful for people. But let me ask - I don't have a whole scope of `/catalog` endpoint. And I want to ask the next: Do we have already `Timestamp` property for a `module` objects?

**Comment by marpio** [comment]
@5onic 
We could probably use the .info files.
Example:
https://proxy.golang.org/github.com/gomods/athens/@v/v0.4.0.info

**Comment by 5onic** [comment]
@marpio Got it, thanks! also would be cool to hear @arschles thoughts.

and jFYI: I'll start work on it since the next Friday, just got a short vacation 😄

**Comment by marpio** [comment]
@5onic 
Thanks for taking this one!
Just for reference, since we talked about this on slack.
@marwan-at-work provided this link https://github.com/golang/go/issues/32312
Apparently it's the time the module entered the database.
We support multiple storages GCS, S3, Azure Blob Storage etc.
Depending on the storage we will have to obtain the timestamp differently i.e. in case of GCS https://godoc.org/cloud.google.com/go/storage#ObjectAttrs

**Comment by gowdabas** [comment]
@arschles Where are we on this? Are we still looking to change the format? If so, I'd like to take this up.

**Comment by marpio** [comment]
@arschles @gowdabas
What do you think about supporting the proxy.golang.org format?

**Comment by marwan-at-work** [comment]
> What do you think about supporting the proxy.golang.org format?

I believe that was something that was discussed (I thought here, but maybe elsewhere now that I don't see it). 

Also, I assume you mean the `index.golang.org` format? 

I'd love to support that format, but I'm also wondering if it's worth supporting both formats? 
The index/golang format is just an append-log and has the same problem as our v1 catalog format: we can't ask the simple question of "how many versions does module X have?" without having to iterate through the entire database.

**Comment by gowdabas** [comment]
@marwan-at-work by both formats, you mean index/golang format and {mod1:[v1, v2,v3], mod2:[v1. v2]} format right? For the second format, yes, we do have the overhead of iterating through the entire database at once and also handling pagination.
@marpio I think we could support both.  What do you think?

**Comment by marpio** [comment]
@marwan-at-work 

> Also, I assume you mean the index.golang.org format?

Yes!

> I'd love to support that format, but I'm also wondering if it's worth supporting both formats?
> The index/golang format is just an append-log and has the same problem as our v1 catalog format: we can't ask the simple question of "how many versions does module X have?" without having to iterate through the entire database.

What do you mean by _both_ formats? Our current v1 and the index.golang.org? We could probably phase out our over time.
The reason, why I think it would be useful o support the `index.golang.org`, is that I assume the future tooling will use it and so Athens would be compatible.
Not sure how stable it is though so waiting for some time is definitely an option.

**Comment by marwan-at-work** [comment]
@marpio by both formats I meant `index.golang.org` and the "v2" format suggested above. 

The reason I say that is because our proposed v2 format will allow us to quickly get all the versions for a given module 

And I'm happy to wait for index/golang to be stable for sure.
